### ISSUE TITLE
Placement of index for the <mroot> element

### ISSUE THREAD

**Issue by fred-wang** [MathML/MathMLinHTML5#13]
cc'ing @khaledhosny

http://www.mathml-association.org/MathMLinHTML5/S3.html#SS3.SSS3

The MATH table spec describes RadicalDegreeBottomRaisePercent as "Height of the bottom of the radical degree, if such is present, in proportion to the ascender of the radical sign".

The MathMLinHTML5 draft says it gives the position of the baseline of the index, instead of the bottom of the index, so this must be fixed.

However, I'm not sure about the last part: "proportion to the ascender of the radical sign". Gecko uses a proportion of the height=ascender+descender of the radical sign.

**Comment by khaledhosny** [comment]
It will need testing on MS implementation to be sure, but being a percent of the ascender not the full height makes sense, this way you can adjust the ascender of the bigger radical variants to get proper positioning for all sizes, otherwise it can look bad the bigger the radical gets (see Cambria Math radical). It looks like LuaTeX uses the full height not the ascender (which would explain why Cambria’s degrees look bad) and XeTeX does not support it at all (it has no built-in support for placing radical degrees, it is currently done by macros).

**Comment by khaledhosny** [comment]
I just checked Math Input Panel on Winows (which I think uses the same implementation used by MS Office, which I don’t have here) and it seems to be using the full height of the radical glyph despite what the spec is saying and how Cambria Math seems to be designed (it looks bad). I think we now need to double check this against the MS Office implementation and then ask for clarification of the spec (either changing it to reflect the reality, or MS confirms this is a bug).

**Comment by fred-wang** [comment]
Here is how Gecko and WebKit place the root index (using a percentage of the full height and the distance is taken from the bottom of the radical sign). I'm not sure what "good" would mean (indices aligned? index placed just above the radical sign?).

As I understand the MATH spec, Height = distance from the baseline of the base and radical ascender means "the part of the radical sign that goes above that baseline after its top with the radical bar".

cc'ing @davidcarlisle who might be able to test on MS Office.

```
    <div>
      <style scoped="scoped">
        math {
        font-family: Latin Modern Math;
        }
        mroot > mn:last-child {
        background: #faa;
        }
      </style>
      <math dir="rtl">
        <mroot>
          <mrow>
            <mn>1</mn>
            <mo>+</mo>
            <mroot>
              <mrow>
                <mn>2</mn>
                <mo>+</mo>
                <mroot>
                  <mrow>
                    <mn>3</mn>
                    <mo>+</mo>
                    <mroot>
                      <mrow>
                        <mn>4</mn>
                        <mo>+</mo>
                        <mroot>
                          <mrow>
                            <mn>5</mn>
                            <mo>+</mo>
                            <mroot>
                              <mrow>
                                <mn>6</mn>
                                <mo>+</mo>
                                <mroot>
                                  <mrow>
                                    <mi>x</mi>
                                  </mrow>
                                  <mn>9</mn>
                                </mroot>
                              </mrow>
                              <mn>8</mn>
                            </mroot>
                          </mrow>
                          <mn>7</mn>
                        </mroot>
                      </mrow>
                      <mn>6</mn>
                    </mroot>
                  </mrow>
                  <mn>5</mn>
                </mroot>
              </mrow>
              <mn>4</mn>
            </mroot>
          </mrow>
          <mn>3</mn>
        </mroot>
      </math>
    </div>

```

![root-index-placement](https://cloud.githubusercontent.com/assets/567455/12866112/e2e4e6e0-ccc1-11e5-939a-60a32f4c383f.png)

**Comment by fred-wang** [comment]
@physikerwelt Or maybe you can test on MS Office too?

**Comment by khaledhosny** [comment]
It depends on font, It looks good here (the indices are relatively at the same distance from the bottom of the radical sign) because the font increases the size of the bottom part as the sign grows, but  if you try it with Cambria Math it looks like this:
![2016-02-07 09-29-21](https://cloud.githubusercontent.com/assets/93914/12871028/e1fc14f8-cd7d-11e5-95b8-ffff616ea884.png)

Which is very ugly IMO, but is the same as MS implementation.

**Comment by fred-wang** [comment]
(a separate issue is that the radical seems to grow much more than needed for Cambria Math, although I don't know if it's a bug in Gecko or in the font)

So if I understand, you'd like the index to be placed just above "the bottom part of the radical sign". If ascender means "the part of the radical going above the baseline", then I still don't see how we can get constant position for a given radical height, since two bases can have the same height but different baselines ; and the position of the top of radical is constrained by Radical(DisplayStyle)VerticalGap + RadicalRuleThickness. If ascender means "the one of the radical glyph" then it would still have to be defined for GlyphAssembly but I don't really see how natural it is to defined a distance from the bottom as a percentage of the glyph ascent...

**Comment by khaledhosny** [comment]
Ascender (as I expect it) is the part of the glyph above the base line *in the font file* not in the laid out line, and thus it is a fixed value for a given glyph and each radical variant can be drawn in the font file in away that the percentage of the ascender gives the desired placement (which I believe how Cambria Math is designed). For a composite radical, I think the bottom glyph would be used for this calculation.

**Comment by fred-wang** [comment]
OK that makes sense. And "height of the bottom of the radical degree" would be the position with respect to the glyph baseline (bottom glyph baseline for GlyphAssembly)... But then that still sounds a complicated way to adjust the position of the index. We could just drop the RadicalDegreeBottomRaisePercent parameter and directly adjust the glyph baseline to position the index.

**Comment by khaledhosny** [comment]
Sure, that why I think we need clarification from MS/OpenType people, either change the spec or admit that MS implementation is wrong.

**Comment by fred-wang** [comment]
In David's screenshot (private email), the gap between the top of the base and the overbar is not constant so it is not uniquely determined by Radical(DisplayStyle)VerticalGap...

**Comment by davidcarlisle** [comment]
![word2010](https://cloud.githubusercontent.com/assets/1268738/12872085/c5ad7c1c-cd8f-11e5-9631-2e379c833a09.png)

**Comment by davidcarlisle** [comment]
The above was with Word 2010, rendering the mathml after removing dir=rtl

**Comment by fred-wang** [comment]
Thanks, the rtl should not matter here (it was in the original MathML test case I used). Also, the index is still very far away from the bottom part of the radical sign, so I'm not sure that's why Khaled calls "good".

**Comment by wspr** [comment]
Oh, I remember implementing this in TeX for unicode-math and wondering why it looked so bad when I was following the spec — can we petition whoever is in charge of OpenType MATH to change this?! The Cambria Math output demonstrably shows that it's just completely broken.

**Comment by khaledhosny** [comment]
The Cambria Math output is bad, which is what I’m trying to say from the start; if the spec was followed it would have fixed the case of Cambria Math here (I think, I haven’t tried that), but it seems even MS is not following the spec here.

**Comment by wspr** [comment]
@khaledhosny Oh, I see what you're saying now. Apologies for the misunderstanding. It certainly would be good to assess. One thing I'm confused about, though, is that the spec talks about the radical being set at a certain percentage of the height of [whatever]. If they were placing it by simply shifting the baseline, why wouldn't it just be at 100% of whatever they chose?

**Comment by fred-wang** [comment]
> If they were placing it by simply shifting the baseline, why wouldn't it just be at 100% of whatever they chose?

@wspr Yes, that's what I was trying to say in https://github.com/MathML/MathMLinHTML5/issues/13#issuecomment-180992579

**Comment by fred-wang** [comment]
From Murray's comment by private mail, Microsoft Office applies RadicalDegreeBottomRaisePercent to the total height of the radical sign, not its ascender. He did not give much details but the calculation he indicates seems to correspond the "MathML in HTML5" implementation note. I also highlighted to him that Microsoft Office's behavior does not really correspond to what the MATH spec says as well as the bad design of Cambria Math (however, IIUC that font was designed by Tiro Typeworks).

I did not get any further feedback on that from Microsoft's employees since Tuesday, so let's just close this issue. If some people want to continue the discussion on Cambria Math & MATH spec on the appropriate channels, then they are more than welcome.
